### PR TITLE
extend documentation of pending tree

### PR DESCRIPTION
**Pull request by AlexHentschel**
This PR aims to extend and polish the documentation of `PendingTree`. Particularly, I am trying to add more details to address this comment in a related PR: https://github.com/onflow/flow-go/pull/4094#discussion_r1150737746

**GoDoc revisions only; no algorithmic changes**

### CODE REVIEW

File: engine/common/follower/pending_tree/pending_tree.go [hunk 1]
```diff
@@ -63,11 +63,25 @@ func (v *PendingBlockVertex) Parent() (flow.Identifier, uint64) {
 // As soon as a valid fork of certified blocks descending from the latest finalized block is observed,
 // we pass this information to caller. Internally, the mempool utilizes the LevelledForest.
 // PendingTree is NOT safe to use in concurrent environment.
-// NOTE: PendingTree relies on notion of `CertifiedBlock` which is a valid block accompanied by a certifying QC (proving block validity).
-// This works well for consensus follower as it is designed to work with certified blocks. To use this structure for consensus
-// participant we can abstract out CertifiedBlock or replace it with a generic argument that satisfies some contract(returns View, Height, BlockID).
-// With this change this structure can be used by consensus participant for tracking connection to the finalized state even without
-// having QC but relying on payload validation.
+// Note:
+//   - The ability to skip ahead is irrelevant for staked nodes, which continuously follows the chain.
```
Comment (jordanschalm) [inline engine/common/follower/pending_tree/pending_tree.go:67]:
```suggestion
//   - The ability to skip ahead is irrelevant for staked nodes, which continuously follow the chain.
```

File: engine/common/follower/pending_tree/pending_tree.go [hunk 2]
```diff
@@ -63,11 +63,25 @@ func (v *PendingBlockVertex) Parent() (flow.Identifier, uint64) {
...
+//     However, light clients don't necessarily follow the chain block by block. Assume a light client
+//     that knows the EpochCommit event, i.e. the consensus committee authorized to certify blocks. A
+//     staked node can easily ship a proof of finalization for a block within that epoch to such a
+//     light client. This would be much cheaper for the light client than downloading the headers for
+//     all blocks in the epoch.
+//   - The pending tree supports skipping ahead, as this is a more general and simpler algorithm.
+//     Removing the ability to skip ahead would restrict the PendingTree's its domain of potential
```
Comment (jordanschalm) [inline engine/common/follower/pending_tree/pending_tree.go:74]:
```suggestion
//     Removing the ability to skip ahead would restrict the PendingTree's domain of potential
```

File: engine/common/follower/pending_tree/pending_tree.go [hunk 3]
```diff
@@ -81,20 +95,22 @@ func NewPendingTree(finalized *flow.Header) *PendingTree {
...
-//  4. Checks for exceeding byzantine threshold. Only one certified block per view is allowed.
-//  5. Finally, blocks with the lowest height from incoming batch that connect to the finalized state we will
-//     mark all descendants as connected, collect them and return as result of invocation.
+// AddBlocks accepts a batch of certified blocks, adds them to the tree of pending blocks and finds blocks connected to
+// the finalized state.
+//
+// Details:
+// Adding blocks might result in additional blocks now being connected to the latest finalized block. The returned
+// slice contains:
+//  1. the subset of `certifiedBlocks` that are connect to the finalized block
```
Comment (jordanschalm) [inline engine/common/follower/pending_tree/pending_tree.go:104]:
```suggestion
//  1. the subset of `certifiedBlocks` that are newly connected to the finalized block
```

**Review by jordanschalm — APPROVED** [review]

**Review by durkmurder — APPROVED** [review]

**Comment by durkmurder** [comment]
I've thought this PR is for feature branch and accidentally applied suggestions :face_with_spiral_eyes:

**Comment by AlexHentschel** [comment]
bors merge

File: engine/common/follower/pending_tree/pending_tree.go (hunk 3 above)
Comment (AlexHentschel) [reply engine/common/follower/pending_tree/pending_tree.go]: @jordanschalm : to be 100% technically accurate, do you think it is really only the _newly_ connected blocks? 

If you call `AddBlock` repeatedly with duplicated blocks, the blocks might already be connected. Though, in a later call, duplicated blocks would still be returned as part of the  `certifiedBlocks` list as they continue to pass this check: https://github.com/onflow/flow-go/blob/08aad74c26df77a99f5237f0258b11f19252c6bd/engine/common/follower/pending_tree/pending_tree.go#L151.
Therefore, I think that the following would be the most technically accurate: 
```
the subset of `certifiedBlocks` that are newly connected to the finalized block
```

what do you think?

**Comment by AlexHentschel** [comment]
bors merge

File: engine/common/follower/pending_tree/pending_tree.go (hunk 3 above)
Comment (AlexHentschel) [reply engine/common/follower/pending_tree/pending_tree.go]: added clarifying comment in https://github.com/onflow/flow-go/pull/4114/commits/3c3f3deca28e8095d186012e23fb0293b03c4df5

**Pull request merged**